Add a function to convert Fortran strings to C strings

MiCurry opened the pull request:
These commits add a new function to smiolf.F90 that converts Fortran strings to C strings and uses that function to replace individual pieces of SMIOLf code that did such conversions.

MiCurry commented:
@mgduda, this change gave me a little more trouble then I had at first thought, but I think it is correct now, but let me know what you think and if you see any problems with it.

I have put the 'Don't Merge' label on this PR as I have only tested it with GNU and MPICH. I'll remove it once I test with other compilers, which I will do so shortly.

Also, in order to test SMIOLf_f_to_c_string, I had to make the routine public, which as a SMIOLf utility subroutine probably shouldn't be. So before merging we'll need to address that in some way.

mgduda commented:
I don't think the function is actually setting all characters in the C array beyond `len_trim(f_string)` to `c_null_char`. Here's minimal code to demonstrate this:
```fortran
character(len=10) :: str                                                                                                             
character (kind=c_char), allocatable, dimension(:) :: cstring                                                                        
integer :: i                                                                                                                         
                                                                                                                                     
str = 'Foobar'                                                                                                                       
                                                                                                                                     
allocate(cstring(32))                                                                                                                
do i=1,32                                                                                                                            
    cstring(i:i) = '-'                                                                                                               
end do                                                                                                                               
                                                                                                                                     
call SMIOLf_f_to_c_string(str, cstring)                                                                                              
                                                                                                                                     
do i=1,32                                                                                                                            
    write(0,*) i, cstring(i:i)                                                                                                       
end do
```

mgduda commented on src/smiolf.F90:
```diff
@@ -1103,4 +1084,35 @@ function SMIOL_free_decomp(decomp) result(ierr) bind(C, name='SMIOL_free_decomp'
+        implicit none
+        
+        character(len=*), intent(in) :: fstring
+        character(kind=c_char), dimension(*), target, intent(out) :: cstring
```
I don't think the `target` attribute is needed here. Also, if `cstring` is an assumed shape array rather than an assumed size array, we can use, e.g., `size(cstring)` within this routine to perhaps simply the code and to validate the size of the `cstring` array.

MiCurry replied:
Okay, that definitely helps and simplifies things. I was using `mpas_c_to_f_string` as an example, but I see now why an assumed-size array wont work, at least with `size` and learned a bit more about assumed-size arrays, so thanks! I'll change the subroutine so that `cstring` is an assumed-shape array and use `size(cstring)` and I'll push that change.

MiCurry commented:
Okay, @mgduda, I have updated this PR with changes to SMIOLf_f_to_c_string that fills the entire `cstring` variable with `c_null_chars`. I also updated the unit test to pre-fill the C string with a default value.

As well I have tested these commits with PGI and Intel, so I will remove the 'Don't Merge' label.

mgduda commented on smiol_runner.F90:
```diff
@@ -1800,4 +1812,55 @@ function compare_decomps(decomp, comp_list_correct, io_list_correct) result(matc
+
+        do i = 1, len(cstring)
+            if (i <= len_trim(str)) then
+                if (all(cstring(i:i) /= str(i:i))) then
```
I may be missing some subtlety, but why is `all` needed here? It looks like a single character is being compared at a time, right?

MiCurry replied:
Looking into it now, I realize its not needed. I expected that `cstring` needed to be accessed in the same way as a Fortran string (i.e. using str(i:i) to access a single character). In that case, `all` was needed to compare the `cstring(i:i)` and `str(i:i)` as the compiler was seeing `cstring(i:i)` as a range, and not a single value.

However, because `cstring` has the dimension attribute we can access a single value by simply using `cstring(i)`, which removes the need to use `all`. So I'll change this PR to remove `all`.

mgduda commented on src/smiolf.F90:
```diff
@@ -1103,4 +1084,35 @@ function SMIOL_free_decomp(decomp) result(ierr) bind(C, name='SMIOL_free_decomp'
+    !> \details
+    !> Converts a Fortran string to a C null-terminated string. The output argument
+    !> must be large enough to contain the Fortran string, plus at least one
+    !> c_null_char character. Any characters beyond len_string(fstring) + 1 of
```
Should `len_string` be `len_trim`?

mgduda commented on smiol_runner.F90:
```diff
@@ -1800,4 +1812,55 @@ function compare_decomps(decomp, comp_list_correct, io_list_correct) result(matc
+
+        call SMIOLf_f_to_c_string(str, cstring)
+
+        do i = 1, len(cstring)
```
I think `len(cstring)` will always be 1; instead, I think we should use `size(cstring)'.

mgduda commented on smiol_runner.F90:
```diff
@@ -1800,4 +1812,55 @@ function compare_decomps(decomp, comp_list_correct, io_list_correct) result(matc
+
+        do i=1, 30
+            cstring(i) = '-'
+        enddo
```
Since `cstring` is an array, it might be simpler to just use `cstring(:) = '-'` rather than using a loop here.

mgduda commented on smiol_runner.F90:
```diff
@@ -1800,4 +1812,55 @@ function compare_decomps(decomp, comp_list_correct, io_list_correct) result(matc
+        ierrcount = 0
+        
+        write(test_log,'(a)',advance='no') 'Everything OK (SMIOLf_f_to_c_string): '
+        allocate(cstring(30))
```
Although `cstring` will be automatically deallocated when this function returns, it seems more complete to explicitly deallocate `cstring` as soon as it is no longer needed. This is just my opinion, though, and perhaps explicitly deallocating in Fortran helps me to remember to free in C.

MiCurry replied:
You're right, even though Fortran would clean it up, its much better to deallocate it when we don't need it.

mgduda commented on src/smiolf.F90:
```diff
@@ -1103,4 +1084,35 @@ function SMIOL_free_decomp(decomp) result(ierr) bind(C, name='SMIOL_free_decomp'
+    !> \brief Convert a Fortran string to a C null-terminated string
+    !> \details
+    !> Converts a Fortran string to a C null-terminated string. The output argument
+    !> must be large enough to contain the Fortran string, plus at least one
```
What happens if the size of `cstring` is not at least `len_trim(fstring)+1`?

MiCurry commented:
> I had in mind two other considerations that we could discuss:
> 
>     1. Should we add additional test cases for `SMIOLf_f_to_c_string`, e.g., for when `size(cstring) < len_trim(fstring)`?
> 
>     2. We could write `SMIOLf_f_to_c_string` as a function that returns an allocatable array of the required size.

I think having a test for when `size(cstring) < len_trim(fstring)` is a good idea. 

I did some tests with `size(cstring) < len_trim(fstring)` and doing such only copies `size(cstring)` of `fstring` into `cstring`. So `fstring = 'abc'` and `size(cstring) == 2`, `cstring` would be set to `ab` with no c null terminating character.

If we update this subroutine to a function, we could add in a check to make sure size(cstring) > len_trim(fstring) or we could just ensure that we always called this function internally and correct and not make it public. Do you have thoughts on that @mgduda?

And too the second point. That makes sense and seems easier. If SMIOLf_f_to_c_string allocated the array then it would be easier to call. Would the function still copy `fstring` into the allocated `cstring`?

mgduda commented:
Here's what I had in mind for a function that returned a C character array that would always be allocated with the correct size:
```fortran
    function SMIOLf_f_to_c_string(fstring) result(cstring)
                                                                                                                                     
        use iso_c_binding, only : c_char, c_null_char                                                                                
                                                                                                                                     
        implicit none                                                                                                                
                                                                                                                                     
        ! Arguments                                                                                                                  
        character(len=*), intent(in) :: fstring                                     
                                                                                                                                     
        ! Return value                                                                                                               
        character(kind=c_char), dimension(:), allocatable :: cstring                                                                 
                                                                                                                                     
        ! Local variables                                                                                                            
        integer :: i                                                                                                                 
                                                                                                                                     
        allocate(cstring(len_trim(fstring)+1))
                                                                                                                                     
        do i = 1, len_trim(fstring)                                                                                                  
            cstring(i) = fstring(i:i)                                                                                                
        end do                                                                                                                       
        cstring(i) = c_null_char                                                                                                     
                                                                                                                                     
    end function SMIOLf_f_to_c_string
```
Annoyingly, though, the GNU compilers generate many warnings like these:
```
smiolf.F90:632:0:

  632 |         c_varname = SMIOLf_f_to_c_string(varname)
      | 
Warning: ‘c_varname.offset’ may be used uninitialized in this function [-Wmaybe-uninitialized]
smiolf.F90:632:0: Warning: ‘c_varname.dim[0].lbound’ may be used uninitialized in this function [-Wmaybe-uninitialized]
smiolf.F90:632:0: Warning: ‘c_varname.dim[0].ubound’ may be used uninitialized in this function [-Wmaybe-uninitialized]
```
This seems to be a GNU issue, and it could be resolved by disabling the `maybe-uninitialized` warning, but this seems like a useful warning to have in general.

MiCurry commented:
@mgduda Which GNU version produced those warnings? I just implemented that change to allocate and return `cstring` and didn't get any warnings with GNU 9.1.0 and GNU 8.3.0.

mgduda commented:
I get warnings with GNU 9.2.0 and 10.1.0. It could also be that I've done something subtly different (and perhaps wrong?) when I implemented the changes. In case you'd like to test with the GNU 10.1.0 compilers, I think there's a module for them available on Cheyenne.

MiCurry commented:
Oh, my mistake. I was still allocated `cstring` before the call to `SMIOLf_f_to_c_string, but now I see the warnings.

Shall we just keep the implementation as it is now in this PR (i.e. have the caller allocate `cstring`)? I think updating the `do loop` to loop from `1` to `len_trim(fstring)` is a good idea.

mgduda commented:
The volume of warning messages from the GNU compiler is annoyingly large, so for now, perhaps the easiest route would be to keep `SMIOLf_f_to_c_string` as a subroutine. But, let's pursue the other clean-up to this PR.

MiCurry commented:
I have just pushed an update to this branch. I have left `SMIOLf_f_to_c_string` as a subroutine and I have added a unit test to test if len(cstring) < len(fstring). I have also left the loop to set cstring as it was originally:

```Fortran
        do i=1, size(cstring)
            cstring(i) = c_null_char
            if (i <= len_trim(fstring)) then
                cstring(i) = fstring(i:i)
            endif
        enddo
```

Looping over `len(fstring)`, when `len(cstring) < len(fstring)` caused a segmentation fault, with the above loop, only `len(cstring)` is copied from `fstring`. The function comment has been updated to reflect this. Another way to implement this subroutine would be to loop over `len(fstring)` there would just need to be a check to see if `cstring` had enough space to hold `fstring + 1` and return an error if it didn't.

mgduda commented:
@MiCurry Other than the unit-test failure with Flang noted above, I've tested successfully with the PGI, GNU, Intel, and XL compilers.

MiCurry commented:
@mgduda Looks good! It looks like Github isn't allowing me to approve my pull request as I am the owner. I think that's because we turned on the feature that each PR needs a single approval before being merged. Should I turn that feature off, or do you want to go ahead an approve it and merge it?

mgduda commented:
@MiCurry I've approved the PR, so I'll try to merge it later today.
